sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/subscriptions/PubSubTestModels.java

```diff
@@ -195,14 +195,58 @@ public Action.Effect<Message> messageOne(@RequestBody Message message) {
 
   @Table(value = "employee_table")
   @Subscribe.EventSourcedEntity(value = EmployeeEntity.class, ignoreUnkown = true)
-  public static class SubscribeOnTypeToEventSourcedEventsWithMethodWithState extends View<Employee> {
+  public static class SubscribeOnTypeToEventSourcedEvents extends View<Employee> {
 
-    public UpdateEffect<Employee> onCreate(Employee employee, EmployeeCreated evt) {
+    public UpdateEffect<Employee> onCreate(EmployeeCreated evt) {
       return effects()
               .updateState(new Employee(evt.firstName, evt.lastName, evt.email));
     }
 
-    public UpdateEffect<Employee> onEmailUpdate(Employee employee, EmployeeEmailUpdated eeu) {
+    public UpdateEffect<Employee> onEmailUpdate(EmployeeEmailUpdated eeu) {
+      var employee = viewState();
+      return effects().updateState(new Employee(employee.firstName, employee.lastName, eeu.email));
+    }
+
+    @Query("SELECT * FROM employees_view WHERE email = :email")
+    @PostMapping("/employees/by-email/{email}")
+    public Employee getEmployeeByEmail(@PathVariable String email) {
+      return null;
+    }
+  }
+
+
+  @Publish.EventStream(source = EmployeeEntity.class, name = "employee_events", ignoreUnknown = true)
+  public static class EventStreamPublishingAction extends Action {
+
+    public String transform(EmployeeCreated created) {
+      return created.toString();
+    }
+  }
```

Comment:
I think this is the best place to show how it would look like if we have separated annotations. 

My original design was something similar to that.

```java
@Subscribe.EventSourcedEntity(value = EmployeeEntity.class, ignoreUnkown = true)
@Publish.EventStream("employee_events")
public static class EventStreamPublishingAction extends Action {

  public String transform(EmployeeCreated created) {
    return created.toString();
  }
}
```

What I don't like with that solution is that the use case is: "User wants to expose the event from entity A to other services". 
But to do that, they need to understand that they need to compose two annotations. I think we do a better job if we provide a single annotation that unambiguously describe the use case. 

And there are some flaws...

* It's possible to only add `@Publish.EventStream("employee_events")` without the `@Subscribe`. Only possible to validate at runtime. Arguably, codegen and annotations processors could validate at compile time, but puts the burden on us to create such validation
* It's possible to combine `@Publish.EventStream("employee_events")` with `@Subscribe.ValueEntity`. Same validation issues as previous point.

In both cases, we will need to add validation to ensure that users are using and combining the annotations in a way that makes sense. While providing one single annotation free us from the task of needing to validate and it frees the user from the task to understand how to compose annotations.

Reply:
In case I would like to publish events to a kafka topic I still need to use 2 annotations right? Or are you planning to merge it all to sth like:
```
@Publish.EventStream(source = EmployeeEntity.class, destination = @Stream("employee_events"), ignoreUnknown = true)
@Publish.EventStream(source = EmployeeEntity.class, destination = @Topic("employee_events"), ignoreUnknown = true)
```

Reply:
I had this thought as well, I was thinking about:

```java
@Publish.Topic(source = EmployeeEntity.class, ...)
```

Source would be optional though because it should be possible to have a REST endpoint that publishes to a Topic. In such a case, the source is the endpoint.

We can explore some ideas. Your proposal looks good as well. The only thing is that will have two different ways to publish to topic (`@Publish.Topic` and `@Publish.EventStream(destination = @Topic)`).

But I absolutely agree that publishing from entity to topic should be done with a single annotation. The question is how.

Reply:
@aludwiko Eventing out to multiple destinations does not work anyway (method eventing.out is a oneof on method level, and on service level only direct is supported - mixing service and method outs leads to validation failure).

I liked it better with the composed/more aligned with the underlying, one reason is that if we diverge completely with the underlying annotations it will be really hard for us to understand/remember how things work. Not sure how strong that argument is though, when it has already diverged with the publish/subscribe vs eventing in/out.

Reply:
> ''What I don't like with that solution is that the use case is: "User wants to expose the event from entity A to other services". But to do that, they need to understand that they need to compose two annotations. I think we do a better job if we provide a single annotation that unambiguously describe the use case.

If that's the main use case, should we also allow the user to define the stream when defining the Entity itself? Rather than needing to define a new action externally? This could be useful when the user is not interested in doing any transformation.

That said, giving the option to change your events for public consumption seems critical to me, and there we would still require the user to have a separate action. Then I think once the user needs to define a new action, it becomes "required" for them to understand this is a two-step thing, so requiring two annotations does not seem farfetched to me. 

> * It's possible to only add `@Publish.EventStream("employee_events")` without the `@Subscribe`. Only possible to validate at runtime. Arguably, codegen and annotations processors could validate at compile time, but puts the burden on us to create such validation

This would need to be prevented for Actions but could be used as a valid scenario for Entities where you don't want to transform the events.

In general, I think I would probably lean towards 2 separate annotations because they seem more composable and closer to what we have at the moment and they seem valuable as well when used separately but I think you're right that validation gets tricker.

Reply:
Note that the publish only without any subscribe will lead to validation failure at discovery, so even if it's not a short feedback loop, it is a feedback loop.

Reply:
I'm not convinced that it's the right thing to have this higher level definition rather than composing in and out. I haven't studied the other Spring annotations for eventing.in (from topic, from event sourced entity, from value entity, from service) and eventing.out (to action, to view, to service).

If we go with these higher level definitions we should do it for all of the combinations, and that could become an explosion.

I was quite happy with that the service-to-service eventing could hook into the same eventing.in and eventing.out concepts as the other things. The code and proto is almost identical when switching between service-to-service and same-service.

Reply:
There are a few reasons why the annotations are not aligned. 

It started with an attempt to use names that are more common or even correct english. There is no such a verb as "to event". The word "eventing" is a made-up concept that needs to be explained or forced into someone else brain. Sure, it's been used by others as well, but stays a strange word in any case. Certainly for newcomers. 

Another reason is that it gives us more type-safety. A generic `@Eventing.In` annotation will require us to have string fields for `topic`, `eventSourcedEntity` and `valueEntity`. 

Specific annotations bounded to some types allows us to inspect the types and pick extra information from there, for example: we can pick the `@EntityType` value and correctly configure the subscription.  

Another reason is that Java annotations, although limited in power, are much more expressive than protobuf annotations. Using protobuf as the common denominator produces a limited Java experience. We see that in other parts as well, for example the fact that event handlers can be typed. 

On the topic of using high-level annotations or not, I can see the concern. But I'm also not convinced that we should let users pick things from the shelf and compose. That means that we put the burden on then to understand what can be composed or not. And we put the burden on us to validate that what they composed is legal or not. 

The more type-safety we have the better. Less validation on our side, more guidance for the users. But true, it means that we need to do more translation on our side. But isn't that the goal of Kalix? Users do less and we do the hard lifting?

Providing high composability looks more something we want for Akka, but in Kalix we want guard-rails.